src/tools/tests/schema2kotlin-test.ts

```diff
@@ -232,27 +263,180 @@ describe('schema2kotlin', () => {
+        'typealias T_H3 = AbstractT.TInternal2',
+        'typealias T_H4 = AbstractT.TInternal2',
+      ]
+    ));
     async function assertSchemaAliases(manifest: string, expectedAliases: string[]) {
       await assertComponent(manifest, ({typeAliases}) => typeAliases.sort(), expectedAliases);
     }
   });
+  describe('Entity Class Generation', () => {
+    it('generates entity with public constructor', async () => await assertClassDefinition(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `@Suppress("UNCHECKED_CAST")
+    class Thing(
+        num: Double = 0.0,
+        entityId: String? = null,
+        creationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
+        expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP
+    ) : EntityBase("Thing", SCHEMA, entityId, creationTimestamp, expirationTimestamp)`
+    ));
+    it('generates variable entity with private constructor', async () => await assertClassDefinition(
+      `particle T
+         h1: reads ~a with {num: Number}
+         `,
+      `@Suppress("UNCHECKED_CAST")
+    class T_H1 private constructor(
+        num: Double = 0.0,
+        entityId: String? = null,
+        creationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
+        expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP
+    ) : VariableEntityBase("T_H1", SCHEMA, entityId, creationTimestamp, expirationTimestamp)`
+    ));
+    it('generates copy and mutate by entity fields', async () => await assertCopyMethods(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `/**
+         * Use this method to create a new, distinctly identified copy of the entity.
+         * Storing the copy will result in a new copy of the data being stored.
+         */
+        fun copy(num: Double = this.num) = Thing(num = num)
+        /**
+         * Use this method to create a new version of an existing entity.
+         * Storing the mutation will overwrite the existing entity in the set, if it exists.
+         */
+        fun mutate(num: Double = this.num) = Thing(
+            num = num,
+            entityId = entityId,
+            creationTimestamp = creationTimestamp,
+            expirationTimestamp = expirationTimestamp
+        )`
+    ));
+    // TODO(alxr): Why do we omit the docstring for Wasm?
+    it('generates only copy method by entity fields for wasm', async () => await assertCopyMethodsForWasm(
+      `particle T
+         h1: reads Thing {num: Number}`,
+      `
+        fun copy(num: Double = this.num) = Thing(num = num)
+        `
+    ));
+    it('copies underlying data for type variables.', async () => await assertCopyMethods(
+      `particle T
+         h1: reads ~a with Thing {num: Number}`,
+      `/**
+         * Use this method to create a new, distinctly identified copy of the entity.
+         * Storing the copy will result in a new copy of the data being stored.
+         */
+        fun copy(num: Double = this.num) = Thing(num = num)
```

Comment:
This makes me think: We did not do anything in VariableEntityBase to copy over the rawSingletons, rawCollections fields on copy() and mutate()

Reply:
I don't think we need to -- they are passed in via the constructor here. i.e., they are in the "other" dictionary.

Reply:
Maybe I don't see something - but copy() simply returns the new instance. The new instance will have new rawSingletons and rawCollections map - we don't seem to copy values from these maps over to the copy.